File: src/falcon_core/instrument_interfaces/physics_units/common_units.py
"""Factory for creating common physical units."""

from .constants import (
    # Conversion factors
    CELSIUS_OFFSET,
    DIMENSIONS_AMPERE,
    DIMENSIONS_CANDELA,
    DIMENSIONS_COULOMB,
    DIMENSIONS_DIMENSIONLESS,
    DIMENSIONS_FARAD,
    DIMENSIONS_HENRY,
    DIMENSIONS_HERTZ,
    DIMENSIONS_JOULE,
    DIMENSIONS_KELVIN,
    DIMENSIONS_KILOGRAM,
    # Dimension dictionaries
    DIMENSIONS_METER,
    DIMENSIONS_MOLE,
    DIMENSIONS_NEWTON,
    DIMENSIONS_OHM,
    DIMENSIONS_PASCAL,
    DIMENSIONS_SECOND,
    DIMENSIONS_SIEMENS,
    DIMENSIONS_TESLA,
    DIMENSIONS_VOLT,
    DIMENSIONS_WATT,
    DIMENSIONS_WEBER,
    ELECTRON_CHARGE,
    FAHRENHEIT_OFFSET,
    KILO_SYMBOL,
    MEGA_SYMBOL,
    MICRO_SYMBOL,
    MILLI_SYMBOL,
    MINUTES_PER_HOUR,
    SECONDS_PER_MINUTE,
    UNIT_SCALE_FAHRENHEIT,
)
from .unit import Unit


class CommonUnits:
    """Factory for common physical units."""

    # SI base units
    METER = Unit(dimensions=DIMENSIONS_METER)
    KILOGRAM = Unit(dimensions=DIMENSIONS_KILOGRAM)
    SECOND = Unit(dimensions=DIMENSIONS_SECOND)
    AMPERE = Unit(dimensions=DIMENSIONS_AMPERE)
    KELVIN = Unit(dimensions=DIMENSIONS_KELVIN)
    MOLE = Unit(dimensions=DIMENSIONS_MOLE)
    CANDELA = Unit(dimensions=DIMENSIONS_CANDELA)

    # Derived SI units
    HERTZ = Unit(dimensions=DIMENSIONS_HERTZ)
    NEWTON = Unit(dimensions=DIMENSIONS_NEWTON)
    PASCAL = Unit(dimensions=DIMENSIONS_PASCAL)
    JOULE = Unit(dimensions=DIMENSIONS_JOULE)
    WATT = Unit(dimensions=DIMENSIONS_WATT)
    COULOMB = Unit(dimensions=DIMENSIONS_COULOMB)
    VOLT = Unit(dimensions=DIMENSIONS_VOLT)
    FARAD = Unit(dimensions=DIMENSIONS_FARAD)
    OHM = Unit(dimensions=DIMENSIONS_OHM)
    SIEMENS = Unit(dimensions=DIMENSIONS_SIEMENS)
    WEBER = Unit(dimensions=DIMENSIONS_WEBER)
    TESLA = Unit(dimensions=DIMENSIONS_TESLA)
    HENRY = Unit(dimensions=DIMENSIONS_HENRY)

    # Common non-SI units
    MINUTE = Unit(
        dimensions=DIMENSIONS_SECOND,  # Same dimensions as second
        scale_factor=SECONDS_PER_MINUTE,
    )
    HOUR = Unit(
        dimensions=DIMENSIONS_SECOND,  # Same dimensions as second
        scale_factor=SECONDS_PER_MINUTE * MINUTES_PER_HOUR,
    )
    ELECTRON_VOLT = Unit(
        dimensions=DIMENSIONS_JOULE,  # Same dimensions as joule
        scale_factor=ELECTRON_CHARGE,
    )

    # Temperature units with offsets
    CELSIUS = Unit(
        dimensions=DIMENSIONS_KELVIN,  # Same dimensions as kelvin
        offset=CELSIUS_OFFSET,
    )
    FAHRENHEIT = Unit(
        dimensions=DIMENSIONS_KELVIN,  # Same dimensions as kelvin
        scale_factor=UNIT_SCALE_FAHRENHEIT,
        offset=FAHRENHEIT_OFFSET,
    )

    # Dimensionless units
    DIMENSIONLESS = Unit(dimensions=DIMENSIONS_DIMENSIONLESS)
    PERCENT = Unit(
        dimensions=DIMENSIONS_DIMENSIONLESS,
        scale_factor=0.01,
    )
    RADIAN = Unit(dimensions=DIMENSIONS_DIMENSIONLESS)  # Dimensionless angle

    # Keep the existing methods unchanged
    @classmethod
    def get_unit_with_prefix(cls, unit: Unit, prefix: str) -> Unit:
        """Get a unit with an SI prefix applied."""
        return unit.with_prefix(prefix)

    @classmethod
    def get_milli(cls, unit: Unit) -> Unit:
        """Get the milli- version of a unit."""
        return cls.get_unit_with_prefix(unit, MILLI_SYMBOL)

    @classmethod
    def get_micro(cls, unit: Unit) -> Unit:
        """Get the micro- version of a unit."""
        return cls.get_unit_with_prefix(unit, MICRO_SYMBOL)

    @classmethod
    def get_kilo(cls, unit: Unit) -> Unit:
        """Get the kilo- version of a unit."""
        return cls.get_unit_with_prefix(unit, KILO_SYMBOL)

    @classmethod
    def get_mega(cls, unit: Unit) -> Unit:
        """Get the mega- version of a unit."""
        return cls.get_unit_with_prefix(unit, MEGA_SYMBOL)

File: src/physics/units/CommonUnits.cpp
#include "falcon_core/physics/units/CommonUnits.hpp"

#include "falcon_core/macros.hpp"
#define SPU(...) SP(Unit, __VA_ARGS__)
namespace falcon_core::physics::units {
const UnitSP CommonUnits::Meter =
    SPU(Unit(SI::DIMENSIONS_METER, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Kilogram =
    SPU(Unit(SI::DIMENSIONS_KILOGRAM, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Second =
    SPU(Unit(SI::DIMENSIONS_SECOND, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Ampere =
    SPU(Unit(SI::DIMENSIONS_AMPERE, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Kelvin =
    SPU(Unit(SI::DIMENSIONS_KELVIN, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Mole =
    SPU(Unit(SI::DIMENSIONS_MOLE, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Candela =
    SPU(Unit(SI::DIMENSIONS_CANDELA, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Hertz =
    SPU(Unit(SI::DIMENSIONS_HERTZ, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Newton =
    SPU(Unit(SI::DIMENSIONS_NEWTON, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Pascal =
    SPU(Unit(SI::DIMENSIONS_PASCAL, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Joule =
    SPU(Unit(SI::DIMENSIONS_JOULE, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Watt =
    SPU(Unit(SI::DIMENSIONS_WATT, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Coulomb =
    SPU(Unit(SI::DIMENSIONS_COULOMB, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Volt =
    SPU(Unit(SI::DIMENSIONS_VOLT, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Farad =
    SPU(Unit(SI::DIMENSIONS_FARAD, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Ohm =
    SPU(Unit(SI::DIMENSIONS_OHM, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Siemens =
    SPU(Unit(SI::DIMENSIONS_SIEMENS, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Weber =
    SPU(Unit(SI::DIMENSIONS_WEBER, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Tesla =
    SPU(Unit(SI::DIMENSIONS_TESLA, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Henry =
    SPU(Unit(SI::DIMENSIONS_HENRY, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Minute =
    SPU(Unit(SI::DIMENSIONS_SECOND, SECONDS_PER_MINUTE, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Hour = SPU(Unit(SI::DIMENSIONS_SECOND,
                                          SECONDS_PER_MINUTE* MINUTES_PER_HOUR,
                                          0.0,
                                          SI::UNIT_SYMBOL));
const UnitSP CommonUnits::ElectronVolt =
    SPU(Unit(SI::DIMENSIONS_JOULE, ELECTRON_CHARGE, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Celsius =
    SPU(Unit(SI::DIMENSIONS_KELVIN, 1.0, CELSIUS_OFFSET, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Fahrenheit = SPU(Unit(SI::DIMENSIONS_KELVIN,
                                                UNIT_SCALE_FAHRENHEIT,
                                                FAHRENHEIT_OFFSET,
                                                SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Dimensionless =
    SPU(Unit(SI::DIMENSIONS_DIMENSIONLESS));
const UnitSP CommonUnits::Percent =
    SPU(Unit(SI::DIMENSIONS_DIMENSIONLESS, 0.01, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::Radian =
    SPU(Unit(SI::DIMENSIONS_DIMENSIONLESS, 1.0, 0.0, SI::UNIT_SYMBOL));
const UnitSP CommonUnits::get_unit_with_prefix(const UnitSP&     unit,
                                               const std::string prefix) {
  return unit->with_prefix(prefix);
}
const UnitSP CommonUnits::get_milli(const UnitSP& unit) {
  return unit->with_prefix(SI::MILLI_SYMBOL);
}
const UnitSP CommonUnits::get_micro(const UnitSP& unit) {
  return unit->with_prefix(SI::MICRO_SYMBOL);
}
const UnitSP CommonUnits::get_nano(const UnitSP& unit) {
  return unit->with_prefix(SI::NANO_SYMBOL);
}
const UnitSP CommonUnits::get_pico(const UnitSP& unit) {
  return unit->with_prefix(SI::PICO_SYMBOL);
}
const UnitSP CommonUnits::get_kilo(const UnitSP& unit) {
  return unit->with_prefix(SI::KILO_SYMBOL);
}
const UnitSP CommonUnits::get_mega(const UnitSP& unit) {
  return unit->with_prefix(SI::MEGA_SYMBOL);
}
const UnitSP CommonUnits::get_giga(const UnitSP& unit) {
  return unit->with_prefix(SI::GIGA_SYMBOL);
}
}  // namespace falcon_core::physics::units
